R/join_keys.R

```diff
@@ -0,0 +1,651 @@
+  }
+
+  checkmate::assert(
+    combine = "or",
+    checkmate::check_character(value),
+    checkmate::check_list(value, names = "named", types = "character", null.ok = TRUE)
+  )
+
+  # Assume characters as being primary keys
+  if (checkmate::test_character(value)) {
+    value <- lapply(dataset_1, function(dataset_ix) {
+      value
+    })
+    names(value) <- dataset_1
+  }
+
+  original_value <- value
+  for (dataset_ix in dataset_1) {
+    if (is.null(value)) {
+      inner_items <- names(join_keys_obj[[dataset_ix]])
+      value <- structure(
+        vector(mode = "list", length = length(inner_items)),
+        names = inner_items
+      )
+    }
+
+    for (new_ix in names(value)) {
+      join_keys_obj[[dataset_ix]][[new_ix]] <- value[[new_ix]]
+    }
+    value <- original_value
+  }
+
+  join_keys_obj
+}
+
+#' @rdname join_keys
+#' @export
+#' @examples
+#'
+#' jk <- join_keys()
+#' jk[["ds1"]] <- list()
+#' jk[["ds2"]][["ds3"]] <- "key"
+#'
+#' jk <- join_keys()
+#' jk[["ds1"]] <- list()
+#' jk[["ds2"]][["ds3"]] <- "key"
+#' jk[["ds4"]] <- list(ds5 = "new")
+#'
+#' jk <- join_keys()
+#' jk[["ds2"]][["ds3"]] <- "key"
+#' jk[["ds2"]][["ds3"]] <- NULL
+#' jk
+`[[<-.join_keys` <- function(join_keys_obj, dataset_1, value) {
+  if (checkmate::test_integerish(dataset_1) || checkmate::test_logical(dataset_1)) {
+    dataset_1 <- names(join_keys_obj)[[dataset_1]]
+  }
+
+  checkmate::assert_string(dataset_1)
+
+  # Accepting 1 subscript with valid `value` formal
+  checkmate::assert_list(value, names = "named", types = "character", null.ok = TRUE)
+
+  # Normalize values
+  norm_value <- lapply(names(value), function(.x) {
+    get_keys(join_key(dataset_1, .x, value[[.x]]))
+  })
+
+  names(norm_value) <- names(value)
+  value <- norm_value
+
+  #
+  # Remove classes to use list-based get/assign operations
+  join_keys_obj <- unclass(join_keys_obj)
+
+  # In case a pair is removed, also remove the symmetric pair
+  removed_names <- setdiff(names(join_keys_obj[[dataset_1]]), names(value))
+  if (length(removed_names) > 0) {
+    for (.x in removed_names) join_keys_obj[[.x]][[dataset_1]] <- NULL
+  }
+
+  join_keys_obj[[dataset_1]] <- value
+
+  # Iterate on all new values to create symmetrical pair
+  for (ds2 in names(value)) {
+    if (ds2 == dataset_1) next
+
+    keep_value <- join_keys_obj[[ds2]] %||% list()
+    new_value <- value[[ds2]]
+
+    if (checkmate::test_character(new_value, min.len = 1, names = "unnamed")) {
+      new_value <- setNames(new_value, new_value)
+    } else if (
+      checkmate::test_character(new_value, min.len = 1)
+    ) {
+      # Invert key
+      new_value <- setNames(names(new_value), new_value)
+    }
+
+    keep_value[[dataset_1]] <- new_value
+
+    # Assign symmetrical
+    join_keys_obj[[ds2]] <- keep_value
+  }
+
+  # Remove NULL or empty keys
+  empty_ix <- vapply(
+    join_keys_obj,
+    function(.x) is.null(.x) || length(.x) == 0,
+    logical(1)
+  )
```

Comment:
`NULL` also has length of 0 so 
`function(.x) is.null(.x) || length(.x) == 0` is the same as `function(.x) length(.x) == 0`
and that means you should be able to do 
`join_keys_obj <- Filter(function(x) length(x) == 0L, join_keys_obj)`

Reply:
Good point! Done